server/src/main/scala/org/http4s/armeria/server/ArmeriaHttp4sHandler.scala

```diff
@@ -133,8 +133,11 @@ private[armeria] class ArmeriaHttp4sHandler[F[_]](
         if (tail == Stream.empty)
           Pull.done
         else
-          Pull.eval(F.async_[Unit] { cb =>
-            discardReturn(writer.whenConsumed().thenRun(() => cb(RightUnit)))
+          Pull.eval(F.async[Unit] { cb =>
+            F.delay(discardReturn(writer.whenConsumed().thenRun(() => cb(RightUnit))))
+              .as(Some(F.delay(
+                writer.close()
+              )))
```

Comment:
Actually doesn't `whenConsumed` also return a `CompletableFuture`? But cancellation may not be working correctly for that 🤔 

Still, you could do like this, not sure if it's better.
```scala
import cats.effect.syntax.all._
F.fromCompletableFuture(F.delay(writer.whenConsumed())).cancelable(F.delay(writer.close()))
```

Reply:
Oh, or maybe even just this.
```scala
import cats.effect.syntax.all._
F.fromCompletableFuture(F.delay(writer.whenConsumed())).onCancel(F.delay(writer.close()))
```

Reply:
> Actually doesn't whenConsumed also return a CompletableFuture? 

yeah, it does

> But cancellation may not be working correctly for that

after glancing at this one more time, I'm not even sure now if cancelation has worked previously. So, good call, Arman. Let's do our best on this

Reply:
> I'm not even sure now if cancelation has worked previously.

The best thing is if we can add some tests for this. Thanks to this semantic change I've been adding tests everywhere and usually discovering that it was already broken 😅 

- https://github.com/http4s/http4s-dom/pull/252
- https://github.com/http4s/http4s-jdk-http-client/pull/875

Reply:
That was a great idea to write a test to prove that cancelation actually works. I found a subtle issue with the client.

Reply:
Q: do we need to call `writer.close()` even though `onFinalize` closes it?